src/operations/blas3/gemm_partial_local.hpp

```diff
@@ -0,0 +1,465 @@
+
+    /* Workgroup id m, k and n */
+    const index_t group_count_mn = group_count_m * group_count_n;
+    const index_t kgroup_id = group_id / group_count_mn;
+    const index_t mn_group_id = group_id - kgroup_id * group_count_mn;
+    const index_t ngroup_id = mn_group_id / group_count_m;
+    const index_t mgroup_id = mn_group_id - ngroup_id * group_count_m;
+
+    /* register offsets */
+    const index_t global_m_offset = mgroup_id * tile_size_dim_m;
+    const index_t global_n_offset = ngroup_id * tile_size_dim_n;
+    const index_t global_k_offset = kgroup_id * tile_size_dim_k * num_tiles;
+
+    /* Find out whether we need to check the limits when loading the tiles */
+    const bool check_m_limit = global_m_offset + tile_size_dim_m > m_;
+    const bool check_n_limit = global_n_offset + tile_size_dim_n > n_;
+
+    /* The first tile is pre-loaded before the loop if double buffering is
+     * enabled */
+    if (double_buffer) {
+      extract_input_blocks(local_id, 0, scratch_ptr, rhs_scratch_ptr,
+                           global_m_offset, global_n_offset, global_k_offset,
+                           check_m_limit, check_n_limit);
+    }
+
+    index_t tile_id = 0;
+    /* Loop over all tiles allocated to this particular workgroup size */
+    do {
+      id.barrier(cl::sycl::access::fence_space::local_space);
+
+      // Start loading the next tile
+      index_t next_tile = double_buffer ? (tile_id + 1) : tile_id;
+      const bool tile_nb_check = do_check<double_buffer>(next_tile < num_tiles);
+      if (tile_nb_check) {
+        extract_input_blocks(local_id, next_tile, scratch_ptr, rhs_scratch_ptr,
+                             global_m_offset, global_n_offset, global_k_offset,
+                             check_m_limit, check_n_limit);
+      }
+
+      if (!double_buffer) {
+        id.barrier(cl::sycl::access::fence_space::local_space);
+      }
+
+      // Calculate offsets into the temporary memory.
+      index_t lhs_offset =
+          (double_buffer * (tile_id & 1) * lhs_tile_mem_size) + m_local_id;
+      index_t rhs_offset =
+          (double_buffer * (tile_id & 1) * rhs_tile_mem_size) + n_local_id;
+
+      /* Loop over the values of a single tile */
+      for (index_t k = 0; k < tile_size_dim_k; k++) {
+        auto idx = 0;
+        auto rhs_index = 0;
+#pragma unroll
+        for (index_t wLPTN = 0; wLPTN < work_per_thread_n; wLPTN++) {
+          // load a RHS element from the scratch buffer
+          const element_t privateRhs = rhs_scratch_ptr[rhs_index + rhs_offset];
+
+          index_t lhs_index = 0;
+#pragma unroll
+          for (index_t wLPTM = 0; wLPTM < work_per_thread_m; wLPTM++) {
+            // load a LHS element from the scratch buffer
+            const element_t privateLhs = scratch_ptr[lhs_index + lhs_offset];
+
+            // Perform a manual MAD.
+            private_res[wLPTM + idx] += (privateLhs * privateRhs);
```

Comment:
replace this with `private_res[wLPTM + id] = cl::sycl::mad(privateLhs, privateRhs, private_res[wLPTM + idx]);`

Reply:
Done